[RunAllTests] Fix #3018, #1675, #2931, #1600, and part of #2642: Enable data module in Bazel

BenHenning opened the pull request:
Fix #3018
Fix #1675
Fix #2931
Fix #1600
Fix part of #2642

## Explanation
This PR:
1. Enables the data module in Bazel by re-enabling Moshi (avoiding needing #1600 altogether--we can continue using Moshi in Bazel) using a similar solution as #2979
2. This adds the data module's tests to the Bazel BUILD graph (which required pulling one fix from #2916). Note that without #2916 these tests will be flaky, but given that they are not required yet this should be fine.
3. This fixes all of the Bazel tests being broken that was caused by adding the Retrofit interfaces & GAE classes as dependencies in Bazel without Moshi being enabled
4. It removes TODOs on #1675 (other than ApplicationModule, no existing tests actually needed a new dependency on NetworkModule so this was a trivial fix)
5. This introduces initial Bazel modularization in GAE for #2931
6. This adds '.aswb' to .gitignore for folks that are using Android Studio with Bazel for development (mainly me at this point)
7. Adds some debugging lines to compute_affected_tests.sh since it seems to keep hanging on my PRs when changing Bazel files; hopefully this will help us better dig into why. It needs to go into a PR where the issue has been observed since otherwise a separate PR that makes the change won't help reveal potential issues.
8. Updates to Robolectric 4.4 to work around the issue described here: https://github.com/oppia/oppia-android/pull/3025#issuecomment-810796701. This upgrade only affects Bazel builds, and conveniently lets us also migrate off of our custom fork of robolectric-bazel and use the shared one, instead.

For those interested, the underlying fix was explicitly declaring a Bazel ``java_plugin`` that points to Moshi's annotation processor. Using a combination of _direct_ dependencies & ``export_plugins`` from ``java_library`` we can explicitly trigger the annotation processor to run for certain libraries (I suspect that Dagger does something similar under the surface, but I haven't dug into it). In the Gradle world ``kapt`` seems to do this automatically for us (or maybe Gradle does). Note that this does *not* use kapt, but it's not actually needed since Moshi is a Java annotation processor.

BenHenning committed:
Enable Moshi support in Bazel.

This includes some reorganizing of the data module, but it introduces
support for generating GAE models through Moshi. Corresponding mock
services & tests are still failing, but //data & //:oppia build.

BenHenning committed:
Make mock services build for testing module.

BenHenning committed:
Add working data tests to Bazel build.

This also fixes a regression from an earlier commit that led to the
Moshi objects not actually being generated correctly, so the data
service tests are good proof that Moshi is being used correctly.

BenHenning committed:
Remove TODOs for #1675.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into enable-data-module-in-bazel

Conflicts:
	.bazelrc

BenHenning commented:
@rt4914 @anandwana001 @jcqli @prayutsu @fsharpasharp PTAL for codeowners & per each of the issues you're associated with.

BenHenning commented:
NB: I'll fix the lint & other failures in a follow-up commit momentarily.

BenHenning committed:
Bazel lint fix.

Avoid relying on native java_library.

BenHenning committed:
Fix broken Gradle build.

This includes explicitly specifying okhttp version since Gradle was
previously randomly picking one.

BenHenning committed:
Add debug details for affected tests script.

BenHenning committed:
Lint fixes.

BenHenning committed:
Attempt to force Bazel cache to clear.

This is an attempt to work around a theorized OOM happening when trying
to compute the sibling build graph for changes build files.

BenHenning commented:
Hmm. I can't seem to get Bazel unstuck, so switching to 'RunAllTests' mode. It seems to be stuck on computing the targets affected by Bazel/BUILD file changes. I'm guessing it's an OOM, but I can't be sure.

BenHenning commented:
Reopening to retrigger tests.

prayutsu reviewed (approved):
LGTM, Thanks @BenHenning

BenHenning commented:
Ah, interesting. The failures look real & related to SSL--we might need to fake https for the sake of testing. Bit weird.

rt4914 reviewed (approved):
LGTM for codeowner files.

jcqli commented on testing/testing_test.bzl:
```diff
@@ -1,5 +1,5 @@
 """
-Macros for domain testing tests.
+Macros for testing tests.
```
This is funny to read, does this mean `testing` module?

BenHenning replied:
Yep. :) It is a bit awkward.

jcqli replied:
Maybe "Macros for tests in the testing module." ?

Super small nit but this might also be something to update in the other `.bzl` files too, I noticed the docstrings have the same format and it was only after looking at the other modules that I could understand this sentence😄

jcqli reviewed (approved):
LGTM, thanks! Left a short comment for clarity

fsharpasharp reviewed (approved):
The modularization of gae looks good.

anandwana001 reviewed (approved):
LGTM

BenHenning commented:
Aha. The SSL failure is due to https://github.com/robolectric/robolectric/issues/5115:

```
1) testStoryService_usingFakeJson_deserializationSuccessful(org.oppia.android.data.backends.gae.api.StoryServiceTest)
java.lang.NullPointerException: No password supplied for PKCS#12 KeyStore.
	at org.bouncycastle.jcajce.provider.keystore.pkcs12.PKCS12KeyStoreSpi.engineLoad(Unknown Source)
	at java.base/java.security.KeyStore.load(KeyStore.java:1479)
	at java.base/sun.security.ssl.TrustStoreManager$TrustAnchorManager.loadKeyStore(TrustStoreManager.java:367)
	at java.base/sun.security.ssl.TrustStoreManager$TrustAnchorManager.getTrustedCerts(TrustStoreManager.java:315)
	at java.base/sun.security.ssl.TrustStoreManager.getTrustedCerts(TrustStoreManager.java:59)
	at java.base/sun.security.ssl.TrustManagerFactoryImpl.engineInit(TrustManagerFactoryImpl.java:51)
	at java.base/javax.net.ssl.TrustManagerFactory.init(TrustManagerFactory.java:278)
	at okhttp3.internal.platform.Platform.platformTrustManager(Platform.kt:84)
	at okhttp3.OkHttpClient.<init>(OkHttpClient.kt:219)
	at okhttp3.OkHttpClient$Builder.build(OkHttpClient.kt:955)
	at org.oppia.android.data.backends.gae.api.StoryServiceTest.setUp(StoryServiceTest.kt:35)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:546)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:252)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.base/java.lang.Thread.run(Thread.java:844)
```

Trying to decide whether to try upgrading to Robolectric 4.4. It's also a bit odd that this doesn't repro locally.

BenHenning committed:
Merge branch 'develop' into enable-data-module-in-bazel

Conflicts:
	third_party/versions.bzl

BenHenning committed:
Post-merge fix.

BenHenning committed:
Fix another post-merge failure.

This regressed in a recent PR likely due to other Bazel failures hiding
it in CI.

BenHenning committed:
Upgrade to Robolectric 4.4.

This also moves off of Oppia's fork of robolectric-bazel.

BenHenning commented:
@anandwana001 PTAL and let me know if you find the Robolectric 4.4 upgrade reasonable. I'm not keen on bundling it into this PR, but it seems to be a potential blocker for re-enabling the data module tests that are needed to verify Moshi is working correctly. This is to be verified (waiting on CI results).

Further, I'm also expecting that TopicRevisionFragmentTest might have 2 failures after upgrading to 4.4; waiting for Bazel results to confirm.

anandwana001 commented:
The image size issue in `TopicRevisionFragmentTest` is not a blocker for version update in robolectric.

BenHenning commented:
I went back to a last known good build (e.g. before the Robolectric tests started failing) and it seems like the same issue occurs. The TopicRevisionFragmentTest failure is directly due to upgrading to Robolectric 4.4, so it will need to be addressed before this can be merged.

BenHenning commented:
Interestingly TopicRevisionFragmentTest passes in Gradle when run with Robolectric 4.4.

BenHenning commented:
Aha. TopicRevisionFragmentTest is not using TestImageLoaderModule. I wonder what about Gradle and/or Robolectric 4.4 leads to it actually passing--we should be using the test module, instead.

BenHenning committed:
Switch TopicRevisionFragmentTest to test image loaders.

The test should have been using them before, but this might have been
missed. The test image loaders are necessary in the Robolectric test
environment, and something about Robolectric 4.4 + Bazel specifically
exposed this.

FareesHussain commented:
@BenHenning are you updating to Robolectric 4.4 in this pr, Then do I have to close https://github.com/oppia/oppia-android/pull/3012

also regarding the passing tests, Just to confirm did you do the Gradle sync after updating it to 4.4 cause the downloaded deps won't change unless, if not that might be the reason `TopicRevisionFragmentTest` passes here (In my case after updating to 4.4 I had to do the Gradle sync )

and if the tests are not passing doesn't that mean we need to resolve #3031 before merging as this might cause CI failure in other pull requests

BenHenning commented:
Cool. Looks like just switching the module completely remedies the issue.

FareesHussain commented:
Just to check 
If `TestImageLoaderModule` fixes the tests in `Robolectric` can you confirm if it works for `Espresso` too

BenHenning commented:
@FareesHussain I did do a Gradle sync. Robolectric runs a bit differently on Bazel (for reasons I both know & don't know), so I'm not too concerned about the discrepancy so long as it passes on both.

I am not updating Robolectric in Gradle builds in this PR, only Bazel, so it shouldn't affect #3031.

Will check Espresso momentarily--I want to verify all Bazel tests pass locally before re-running CI since it's much faster to check locally.

BenHenning commented:
Looks like there are no changes in develop to pull in, and I verified all Bazel tests were passing on the previous commit, and the last one only affected 1 test which passed locally.

Confirming that TopicRevisionFragmentTest is also passing via Espresso + Gradle locally.

Going to restart the full CI test suite here since it should pass, and I'm comfortable merging in the latest change without a follow-up review.

BenHenning commented:
Reopening to restart CI.

BenHenning commented:
Actually, I might merge this after Gradle is passing since I'm pretty confident the Bazel issue is resolved (previous test suite fully passed with only the most recently fixed test failing, and it's passing locally now). I'd rather get this in to reduce confusion among contributors who are seeing all the Bazel failures (and to avoid regressions--I had to fix a couple after the latest develop merge).

Edit: I'll fix any issues I missed on the Bazel side in a follow-up PR if needed since I don't expect there to be any now.

FareesHussain commented:
@BenHenning as [these tests](https://github.com/oppia/oppia-android/pull/3012#issuecomment-807313248) are failing due to `GlideImageLoaderModule` so https://github.com/oppia/oppia-android/pull/3012 is ready for review and 
https://github.com/oppia/oppia-android/pull/3032#issuecomment-811076464 is not dependent on the tests.

BenHenning commented:
Ah, that's convenient @FareesHussain. I suspect that you can just wait for this PR to merge & then update #3012 with the changes. Upgrading to 4.4 to fix the 501 issue still seems valuable, and at that point there won't be CI failures anymore which will unblock it.

BenHenning commented:
Ah, @anandwana001 it appears you do need to approve this again since I re-requested your review. PTAL with priority.

BenHenning commented:
Ah. Hmm apparently approval wasn't needed. I still don't understand GitHub codeowners apparently. :)
